crates/metadata-store/proto/metadata_store_svc.proto

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
+//
+// This file is part of the Restate service protocol, which is
+// released under the MIT license.
+//
+// You can find a copy of the license in file LICENSE in the root
+// directory of this repository or package, or at
+// https://github.com/restatedev/proto/blob/main/LICENSE
+
+syntax = "proto3";
+
+import "google/protobuf/empty.proto";
+
+package dev.restate.metadata_store_svc;
+
+// Grpc service definition for a MetadataStore implementation.
+service MetadataStoreSvc {
+  // Get a versioned kv-pair
+  rpc Get(GetRequest) returns (GetResponse);
+
+  // Get the current version for a kv-pair
+  rpc GetVersion(GetRequest) returns (GetVersionResponse);
+
+  // Puts the given kv-pair into the metadata store
+  rpc Put(PutRequest) returns (google.protobuf.Empty);
+
+  // Deletes the given kv-pair
+  rpc Delete(DeleteRequest) returns (google.protobuf.Empty);
+}
+
+message GetRequest {
+  string key = 1;
+}
+
+message PutRequest {
+  string key = 1;
+  VersionedValue value = 2;
+  Precondition precondition = 3;
+}
+
+message DeleteRequest {
+  string key = 1;
+  Precondition precondition = 2;
+}
+
+message VersionedValue {
+  Version version = 1;
+  bytes bytes = 2;
+}
+
+message GetResponse {
+  optional VersionedValue value = 2;
```

Comment:
Not sure if optional here is needed, non-scalar messages are always optional in protof IIRC, but it's okay if you want to keep it to document the expected semantics.

Reply:
Yes, it is mainly for documentation purposes. The index should probably be 1 instead of 2.